na_ucx_am_send_retry.patch

```diff
@@ -0,0 +1,171 @@
+         .op_attr_mask = UCP_OP_ATTR_FIELD_REQUEST | UCP_OP_ATTR_FIELD_CALLBACK |
+-                        UCP_OP_ATTR_FIELD_FLAGS,
++                        UCP_OP_ATTR_FIELD_FLAGS | UCP_OP_ATTR_FIELD_USER_DATA,
+         .cb = {.send = na_ucp_am_send_cb},
+         .flags = UCP_AM_SEND_FLAG_REPLY,
+-        .request = request};
++        .request = request,
++        .user_data = na_ucx_class};
+     ucs_status_ptr_t status_ptr;
+     na_return_t ret;
+ 
+@@ -1809,9 +1820,10 @@ error:
+ 
+ /*---------------------------------------------------------------------------*/
+ static void
+-na_ucp_am_send_cb(void *request, ucs_status_t status, void NA_UNUSED *user_data)
++na_ucp_am_send_cb(void *request, ucs_status_t status, void *user_data)
+ {
+     na_return_t cb_ret;
++    struct na_ucx_class * na_ucx_class = (struct na_ucx_class *) user_data;
+ 
+     NA_LOG_SUBSYS_DEBUG(
+         msg, "ucp_am_send_nbx() completed (%s)", ucs_status_string(status));
+@@ -1820,9 +1832,13 @@ na_ucp_am_send_cb(void *request, ucs_status_t status, void NA_UNUSED *user_data)
+         NA_GOTO_DONE(done, cb_ret, NA_SUCCESS);
+     if (status == UCS_ERR_CANCELED)
+         NA_GOTO_DONE(done, cb_ret, NA_CANCELED);
++    else if (status == UCS_ERR_NO_MEMORY && na_ucx_class->remap_send_nomem)
++        NA_GOTO_SUBSYS_ERROR(msg, done, cb_ret, NA_TIMEOUT,
++            "ucp_am_send_nbx() failed (NA_NOMEM mapped to NA_TIMEOUT)");
+     else
+         NA_GOTO_SUBSYS_ERROR(msg, done, cb_ret, na_ucs_status_to_na(status),
+-            "ucp_am_send_nbx() failed (%s)", ucs_status_string(status));
++            "ucp_am_send_nbx() failed (%s), status: (%d)",
++            ucs_status_string(status), status);
+ 
+ done:
+     na_ucx_complete((struct na_ucx_op_id *) request, cb_ret);
+@@ -2022,6 +2038,9 @@ na_ucp_msg_send_cb(
+         NA_GOTO_DONE(done, cb_ret, NA_SUCCESS);
+     if (status == UCS_ERR_CANCELED)
+         NA_GOTO_DONE(done, cb_ret, NA_CANCELED);
++    else if (status = UCS_ERR_NO_MEMORY)
++        NA_GOTO_SUBSYS_ERROR(msg, done, cb_ret, NA_TIMEOUT,
++            "ucp_am_send_nbx() failed (NA_NOMEM mapped to NA_TIMEOUT)");
+     else
+         NA_GOTO_SUBSYS_ERROR(msg, done, cb_ret, na_ucs_status_to_na(status),
+             "ucp_tag_send_nbx() failed (%s)", ucs_status_string(status));
+@@ -2945,6 +2964,25 @@ na_ucx_release(void *arg)
+     }
+ }
+ 
++/*---------------------------------------------------------------------------*/
++static na_return_t
++na_ucx_class_env_config(struct na_ucx_class * na_ucx_class)
++{
++    char *env;
++    na_return_t ret;
++
++    /* Set unexpected msg callbacks */
++    env = getenv("NA_UCX_REMAP_SEND_NOMEM");
++    if (env != NULL && (env[0] == '1' || tolower(env[0]) == 'y' ||
++        tolower(env[0]) == 't')) {
++        na_ucx_class->remap_send_nomem = true;
++    } else {
++        NA_LOG_SUBSYS_DEBUG(cls, "NA_UCX_REMAP_SEND_NOMEM set");
++        na_ucx_class->remap_send_nomem = false;
++    }
++
```

Comment:
it looks like you're missing the return value here? also the debug message should be "NA_UCX_REMAP_SEND_NOMEM not set" ?

Reply:
I intended to only log when it's set, but Alex and I discussed and made setting it the default.  I've changed to log when it is set (and I fixed the return).